Extensions/Xtensive.Orm.BulkOperations/Internals/QueryOperation.cs

```diff
@@ -37,30 +36,38 @@ protected override int ExecuteInternal()
     {
       Expression e = query.Expression.Visit((MethodCallExpression ex) =>
         {
-          if (ex.Method.DeclaringType == typeof (QueryableExtensions) && ex.Method.Name == "In" &&
-              ex.Arguments.Count > 1)
-          {
-            if (ex.Arguments[1].Type == typeof (IncludeAlgorithm))
-            {
-              var v = (IncludeAlgorithm) ex.Arguments[1].Invoke();
-              if (v == IncludeAlgorithm.TemporaryTable)
-              {
+          var methodInfo = ex.Method;
+          //rewrite localCollection.Contains(entity.SomeField) -> entity.SomeField.In(localCollection)
+          if (methodInfo.DeclaringType == typeof(Enumerable) &&
+              methodInfo.Name == "Contains" &&
+              ex.Arguments.Count == 2) {
+            var localCollection = ex.Arguments[0];//IEnumerable<T>
+            var valueToCheck = ex.Arguments[1];
+            var genericInMethod = inMethod.MakeGenericMethod(new[] { valueToCheck.Type });
+            ex = Expression.Call(genericInMethod, valueToCheck, Expression.Constant(IncludeAlgorithm.ComplexCondition), localCollection);
+            methodInfo = ex.Method;
+          }
+
+          if (methodInfo.DeclaringType == typeof(QueryableExtensions) &&
+              methodInfo.Name == "In" &&

```

Comment:
It is better to use `StringComparison.Ordinal` on a hot paths like expression tree processing.